src/_P112_AS7265x.ino

```diff
@@ -259,9 +253,8 @@ boolean Plugin_112(byte function, struct EventStruct *event, String& string)
           UserVar[event->BaseVarIndex + 2] = P112_data->sensor.getR();
           UserVar[event->BaseVarIndex + 3] = P112_data->sensor.getW();
         }
-      } else if (MeasurementStatus != "Running") {
-        MeasurementStatus = String("Running");
-        eventQueue.add("PLUGIN_READ: MeasurementStatus = " + MeasurementStatus);
+      } else if (P112_data->MeasurementStatus != 1) {
+        P112_data->MeasurementStatus = 1;

```

Comment:
What is state 0, 1 and 2?
Maybe you can use a `#define P112_STATE_MEASUREMENT_ACTIVE  1` to clarify?

Also a check for `P112_data->MeasurementStatus != 1` is also true if `P112_data->MeasurementStatus == 0`
Is that correct?

Reply:
A short summary is already included into P112_data_struct.h:

```
struct P112_data_struct : public PluginTaskData_base {
  bool begin();
  AS7265X sensor;
  bool initialized = false;

  // MeasurementStatus:
  // 0 : Not running
  // 1 : Running
  // 2 : Sensor data available
  byte MeasurementStatus = 0;
};
```

> _A check for P112_data->MeasurementStatus != 1 is true if P112_data->MeasurementStatus == 0
> Is that correct?_

Jepp - that is correct

![image](https://user-images.githubusercontent.com/3251642/113126653-29050f80-9218-11eb-8492-e7c7eeba12db.png)

Reply:
600 msec is still quite long.
Have to see if we can sample per wavelength

Reply:
![image](https://user-images.githubusercontent.com/3251642/113136381-94081380-9223-11eb-9244-ef4969ff0d64.png)

I have implemented mode 3
Mode 2 is running continiously - Makes really no sense - Is also not reducing the reading time

Reply:
I have disabled every sensor reading exept the last one.
Reading time has been reduced to 35 ms:

```
        if (PCONFIG(3)) // Calibrated Measurements?
        {
//          eventQueue.add(RuleEvent + "410=" + P112_data->sensor.getCalibratedA());
//          eventQueue.add(RuleEvent + "435=" + P112_data->sensor.getCalibratedB());
//          eventQueue.add(RuleEvent + "460=" + P112_data->sensor.getCalibratedC());

//          eventQueue.add(RuleEvent + "485=" + P112_data->sensor.getCalibratedD());
//          eventQueue.add(RuleEvent + "510=" + P112_data->sensor.getCalibratedE());
//          eventQueue.add(RuleEvent + "535=" + P112_data->sensor.getCalibratedF());

//          eventQueue.add(RuleEvent + "560=" + P112_data->sensor.getCalibratedG());
//          eventQueue.add(RuleEvent + "585=" + P112_data->sensor.getCalibratedH());
//          eventQueue.add(RuleEvent + "610=" + P112_data->sensor.getCalibratedR());

//          eventQueue.add(RuleEvent + "645=" + P112_data->sensor.getCalibratedI());
//          eventQueue.add(RuleEvent + "680=" + P112_data->sensor.getCalibratedS());
//          eventQueue.add(RuleEvent + "705=" + P112_data->sensor.getCalibratedJ());

//          eventQueue.add(RuleEvent + "730=" + P112_data->sensor.getCalibratedT());
//          eventQueue.add(RuleEvent + "760=" + P112_data->sensor.getCalibratedU());
//          eventQueue.add(RuleEvent + "810=" + P112_data->sensor.getCalibratedV());

//          eventQueue.add(RuleEvent + "860=" + P112_data->sensor.getCalibratedW());
//          eventQueue.add(RuleEvent + "900=" + P112_data->sensor.getCalibratedK());
//          eventQueue.add(RuleEvent + "940=" + P112_data->sensor.getCalibratedL());

//          UserVar[event->BaseVarIndex + 0] = P112_data->sensor.getCalibratedC();
//          UserVar[event->BaseVarIndex + 1] = P112_data->sensor.getCalibratedF();
//          UserVar[event->BaseVarIndex + 2] = P112_data->sensor.getCalibratedR();
          UserVar[event->BaseVarIndex + 3] = P112_data->sensor.getCalibratedW();
        } else {
```

![image](https://user-images.githubusercontent.com/3251642/113138113-a7b47980-9225-11eb-9ba4-5680bf961e02.png)

Reply:
So you need some extra step, to collect the value of each wavelength.
If it is taking roughly 35 msec, I guess you can do it in the `PLUGIN_TEN_PER_SEC` call. (leaving some time left for the other tasks running on the ESP)

Reply:
New try:

![image](https://user-images.githubusercontent.com/3251642/113270205-87de8d80-92d9-11eb-8b8b-6c1a58df79e1.png)

Reply:
You should also look at the PLUGIN_TEN_PER_SECOND as that is the one performing the reads.
78 msec on reading is still quite long, but I don't expect this can be made shorter for this sensor.

Since you're reading them in a burst and probably more often than you would usually do, the PLUGIN_TEN_PER_SECOND call is called less than 7x a second, so you can already see it has quite some impact on the operations of the rest of the node.
For example if you're running some other sensor which constantly delivers data, like a GPS, or some other serial port sensor, then it is very likely you will miss data as buffers will overflow as this plugin is blocking for 70-ish msec each 100 msec, for 18 sensor values.